src/main/java/com/endava/internship/service/UserServiceImpl.java

```diff
@@ -2,68 +2,105 @@
 
 import com.endava.internship.domain.Privilege;
 import com.endava.internship.domain.User;
-import com.endava.internship.service.UserService;
 
+import java.util.Arrays;
+import java.util.Comparator;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.function.Function;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
 
 public class UserServiceImpl implements UserService {
 
     @Override
-    public List<String> getFirstNamesReverseSorted(List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+    public List<String> getFirstNamesReverseSorted(final List<User> users) {
+        return users.stream()
+                .map(User::getFirstName)
+                .sorted(Comparator.reverseOrder())
+                .collect(Collectors.toList());
     }
 
     @Override
     public List<User> sortByAgeDescAndNameAsc(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .sorted(Comparator.comparingInt(User::getAge).reversed()
+                        .thenComparing(User::getFirstName))
+                .collect(Collectors.toList());
     }
 
     @Override
     public List<Privilege> getAllDistinctPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .flatMap(user -> user.getPrivileges().stream())
+                .distinct()
+                .collect(Collectors.toList());
     }
 
     @Override
     public Optional<User> getUpdateUserWithAgeHigherThan(final List<User> users, final int age) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .filter(user -> user.getPrivileges().contains(Privilege.UPDATE) && user.getAge() > age)
+                .findAny();
     }
 
     @Override
     public Map<Integer, List<User>> groupByCountOfPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .collect(Collectors.groupingBy(user -> user.getPrivileges().size()));
     }
 
     @Override
     public double getAverageAgeForUsers(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .mapToDouble(User::getAge)
+                .average().orElse(-1);
     }
 
     @Override
     public Optional<String> getMostFrequentLastName(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        final var map = getNumberOfLastNames(users);
+        final var list = map.entrySet().stream()
```

Comment:
why don't you want to make something like Optional.ofNullable(users).map(::stream)....
why defining a new map? it consumes lot of memory